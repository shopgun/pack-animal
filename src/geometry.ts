import { Matrix } from "./vendor/matrix";
import polygonOverlap from "./vendor/polygon-overlap";

export interface IPoint {
  x: number;
  y: number;
}

export const rotateMatrixAroundPoint = (point: IPoint, degrees: number, matrix = new Matrix()) => {
  const rotatedMatrix = matrix.clone();
  const { x, y } = point;
  rotatedMatrix.multiply(new Matrix().translate(x, y));
  rotatedMatrix.multiply(new Matrix().rotateDeg(degrees));
  rotatedMatrix.multiply(new Matrix().translate(-x, -y));
  return rotatedMatrix;
};

export const doPolygonsOverlap = (polygon1: IPoint[], polygon2: IPoint[]) =>
  polygonOverlap(polygon1.map(({ x, y }) => [x, y]), polygon2.map(({ x, y }) => [x, y]));

export const polygonArea = (points: IPoint[]) => {
  const l = points.length;
  let det = 0;

  points = points.concat(points[0]);

  for (let i = 0; i < l; i++) {
    det += points[i].x * points[i + 1].y - points[i].y * points[i + 1].x;
  }

  return Math.abs(det) / 2;
};

export const polygonBounds = (points: IPoint[]): IPoint[] => {
  const left = Math.min(...points.map(point => point.x));
  const top = Math.min(...points.map(point => point.y));
  const right = Math.max(...points.map(point => point.x));
  const bottom = Math.max(...points.map(point => point.y));
  return [
    { x: left, y: top },
    { x: right, y: top },
    { x: right, y: bottom },
    { x: left, y: bottom }
  ];
};
export const polygonsBounds = (polygons: IPoint[][]): IPoint[] => {
  const polygonsPoints = Array.prototype.concat(...polygons);
  return polygonBounds(polygonsPoints);
};

export const isPolygonWithinRectangle = (polygon: IPoint[], rectangle: IPoint[]) => {
  if (!polygon.length) {
    throw new Error("Checking a pointless polygon is pointless.");
  }
  for (const { x, y } of polygon) {
    if (x < rectangle[0].x || x > rectangle[1].x) {
      return false;
    }
    if (y < rectangle[0].y || y > rectangle[1].y) {
      return false;
    }
  }
  return true;
};

export const verifyPack = (polygons: IPoint[][], rectangle: IPoint[]) =>
  polygons.every(polygon => isPolygonWithinRectangle(polygon, rectangle)) &&
  !(
    polygons.length > 1 &&
    polygons.some((polygon, index) =>
      polygons
        .filter((_, arrIndex) => index !== arrIndex)
        .some(otherPolygon => doPolygonsOverlap(polygon, otherPolygon))
    )
  );

export const packUtilization = (
  rectangleWidth: number,
  rectangleHeight: number,
  polygons: IPoint[][]
) => {
  const rectanglePolygon: IPoint[] = [
    { x: 0, y: 0 },
    { x: rectangleWidth, y: 0 },
    { x: rectangleWidth, y: rectangleHeight },
    { x: 0, y: rectangleHeight }
  ];
  const rectangleArea = polygonArea(rectanglePolygon);
  const utilization =
    1 -
    polygons.reduce((memo, points) => memo - polygonArea(points), rectangleArea) / rectangleArea;
  return utilization;
};

export const radiansToDegrees = (radians: number) => radians * 180 / Math.PI;
