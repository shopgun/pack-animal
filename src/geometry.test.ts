import { packUtilization, polygonArea, polygonBounds } from "./geometry";

describe("geometry: polygonBounds", () => {
  it("gives a bounding box polygon from any polygon", () => {
    expect(polygonBounds([{ x: 0, y: 0 }, { x: 5, y: 0 }, { x: 2.5, y: 5 }])).toEqual([
      { x: 0, y: 0 },
      { x: 5, y: 0 },
      { x: 5, y: 5 },
      { x: 0, y: 5 }
    ]);
  });
});
describe("geometry: polygonArea", () => {
  it("calculates total area from any polygon", () => {
    expect(polygonArea([{ x: 0, y: 0 }, { x: 0, y: 4 }, { x: 4, y: 4 }, { x: 4, y: 0 }])).toEqual(
      16
    );
  });
});
describe("geometry: packUtilization", () => {
  it("calculates area utilization of polygons in rectangle", () => {
    expect(
      packUtilization(800, 742, [
        [
          { x: 562.83, y: 453.34 },
          { x: 504.37, y: 519.2 },
          { x: 512.14, y: 682.37 },
          { x: 575.41, y: 740.46 },
          { x: 739.69, y: 733.06 },
          { x: 799.26, y: 666.8299999999999 },
          { x: 794.8199999999999, y: 509.95 },
          { x: 726, y: 445.57 }
        ],
        [
          { x: 296.37, y: 445.57 },
          { x: 308.21, y: 560.64 },
          { x: 296.37, y: 740.46 },
          { x: 496.53999999999996, y: 740.46 },
          { x: 496.53999999999996, y: 445.57 }
        ],
        [
          { x: 0.37, y: 534.61 },
          { x: 0.37, y: 734.78 },
          { x: 295.26, y: 734.78 },
          { x: 295.26, y: 695.9300000000001 },
          { x: 236.43, y: 718.5 },
          { x: 263.07, y: 681.13 },
          { x: 251.23, y: 636.36 },
          { x: 269.36, y: 593.07 },
          { x: 295.26, y: 677.4300000000001 },
          { x: 295.26, y: 534.61 }
        ],
        [
          { x: 0.37, y: 326.85 },
          { x: 0.37, y: 397.89 },
          { x: 27.009999999999998, y: 394.56 },
          { x: 24.419999999999998, y: 508.89 },
          { x: 0.37, y: 404.55 },
          { x: 0.37, y: 529.98 },
          { x: 295.26, y: 529.98 },
          { x: 295.26, y: 326.85 }
        ],
        [
          { x: 631.3652546097812, y: 104.4224268168694 },
          { x: 629.4777974194814, y: 125.82806171412918 },
          { x: 648.3542831423874, y: 143.05279064531783 },
          { x: 610.0110938204541, y: 171.2912649777315 },
          { x: 582.2043082486331, y: 369.1468251122973 },
          { x: 597.6524825763552, y: 397.47245855517565 },
          { x: 747.8761486044514, y: 418.5850179708175 },
          { x: 768.1350591138255, y: 399.38768625588455 },
          { x: 795.4269042120944, y: 205.19611797566262 },
          { x: 768.4183461622583, y: 155.44305985790834 },
          { x: 774.0094075463616, y: 128.95338955797465 }
        ]
      ])
    ).toMatchSnapshot();
  });
});
