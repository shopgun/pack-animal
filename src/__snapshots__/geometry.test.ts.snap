// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geometry: packUtilization calculates area utilization of polygons in rectangle 1`] = `0.5027201624831534`;
